crates/storage/provider/src/traits/stage_checkpoint.rs

```diff
@@ -0,0 +1,16 @@
+use reth_interfaces::Result;
+use reth_primitives::{stage::StageId, BlockNumber, StageCheckpoint};
+
+/// The trait for fetching stage checkpoint related data.
+#[auto_impl::auto_impl(&, Arc)]
+pub trait StageCheckpointProvider: Send + Sync {
+    /// Fetch the checkpoint for the given stage.
+    fn get_stage_checkpoint(&self, id: StageId) -> Result<Option<StageCheckpoint>>;
+
+    /// Fetch the checkpoint for the last stage ([reth_primitives::stage::StageId::Finish]).
+    /// Should return `0` if no progress has been made yet.
+    ///
+    /// NOTE: This checkpoint is always less than the checkpoint of any other stage. Hence it's used
+    /// to denote the minimum number of the block that was **fully** processed by the pipeline.
+    fn get_last_stage_checkpoint(&self) -> Result<BlockNumber>;
```

Comment:
this is a bit confusing, can we name this `finish` instead?

this can be a default impl that calls `get_stage_checkpoint`?

Reply:
> NOTE: This checkpoint is always less than the checkpoint of any other stage.

I don't think this is necessarily true, because directly after Finish stage everything should be on the __same__ checkpoint.

This essentially represents the fully synced block, that is easier to understand.